src/vehicle_gui.cpp

```diff
@@ -110,30 +144,37 @@ uint GetUnitNumberDigits(VehicleList &vehicles)
 		unitnumber = max<uint>(unitnumber, v->unitnumber);
 	}
 
-	if (unitnumber >= 10000) return 5;
-	if (unitnumber >=  1000) return 4;
-	if (unitnumber >=   100) return 3;
-
-	/*
-	 * When the smallest unit number is less than 10, it is
-	 * quite likely that it will expand to become more than
-	 * 10 quite soon.
-	 */
-	return 2;
+	return CountDigitsForAllocatingSpace(unitnumber);
 }
 
 void BaseVehicleListWindow::BuildVehicleList()
 {
-	if (!this->vehicles.NeedRebuild()) return;
+	if (!this->vehgroups.NeedRebuild()) return;
 
 	DEBUG(misc, 3, "Building vehicle list type %d for company %d given index %d", this->vli.type, this->vli.company, this->vli.index);
 
+	this->vehgroups.clear();
+
+	/* Note: This can be optimized so that merely changing the "group by" dropdown value does not need to call GenerateVehicleSortList again. */
 	GenerateVehicleSortList(&this->vehicles, this->vli);
 
-	this->unitnumber_digits = GetUnitNumberDigits(this->vehicles);
+	uint max_unitnumber = 0;
+	for (auto it = this->vehicles.begin(); it != this->vehicles.end(); ++it) {
+		this->vehgroups.emplace_back();
+		GUIVehicleGroup& vehgroup = this->vehgroups.back();
 
-	this->vehicles.RebuildDone();
-	this->vscroll->SetCount((uint)this->vehicles.size());
+		vehgroup.vehicles_begin = it;
+		vehgroup.vehicles_end = it + 1;
+		vehgroup.display_profit_this_year = (*it)->GetDisplayProfitThisYear();
+		vehgroup.display_profit_last_year = (*it)->GetDisplayProfitLastYear();
+		vehgroup.age = (*it)->age;
```

Comment:
seems like it would be nicer if GUIVehicleGroup had a constructor, so you didn't have to set all these values manually

Reply:
Agreed with this, I've added a constructor.